Source/WebCore/loader/HistoryController.cpp

```diff
@@ -294,40 +294,48 @@ bool HistoryController::shouldStopLoadingForHistoryItem(HistoryItem& targetItem)
 
 // Main funnel for navigating to a previous location (back/forward, non-search snap-back)
 // This includes recursion to handle loading into framesets properly
-void HistoryController::goToItem(HistoryItem& targetItem, FrameLoadType type, ShouldTreatAsContinuingLoad shouldTreatAsContinuingLoad)
+void HistoryController::goToItem(HistoryItem& targetItem, FrameLoadType frameLoadType, ShouldTreatAsContinuingLoad shouldTreatAsContinuingLoad)
 {
-    LOG(History, "HistoryController %p goToItem %p type=%d", this, &targetItem, static_cast<int>(type));
+    LOG(History, "HistoryController %p goToItem %p type=%d", this, &targetItem, static_cast<int>(frameLoadType));
 
-    // shouldGoToHistoryItem is a private delegate method. This is needed to fix:
-    // <rdar://problem/3951283> can view pages from the back/forward cache that should be disallowed by Parental Controls
-    // Ultimately, history item navigations should go through the policy delegate. That's covered in:
-    // <rdar://problem/3979539> back/forward cache navigations should consult policy delegate
     RefPtr page = m_frame->page();
     if (!page)
         return;
-    if (!m_frame->loader().client().shouldGoToHistoryItem(targetItem))
-        return;
-    if (m_defersLoading) {
-        m_deferredItem = &targetItem;
-        m_deferredFrameLoadType = type;
-        return;
-    }
 
-    // Set the BF cursor before commit, which lets the user quickly click back/forward again.
-    // - plus, it only makes sense for the top level of the operation through the frame tree,
-    // as opposed to happening for some/one of the page commits that might happen soon.
-    CheckedRef backForward = page->backForward();
-    RefPtr currentItem = backForward->currentItem(m_frame->frameID());
-    backForward->setProvisionalItem(targetItem);
-
-    // First set the provisional item of any frames that are not actually navigating.
-    // This must be done before trying to navigate the desired frame, because some
-    // navigations can commit immediately (such as about:blank).  We must be sure that
-    // all frames have provisional items set before the commit.
-    recursiveSetProvisionalItem(targetItem, currentItem.get(), ForNavigationAPI::No);
-
-    // Now that all other frames have provisional items, do the actual navigation.
-    recursiveGoToItem(targetItem, currentItem.get(), type, shouldTreatAsContinuingLoad);
+    auto finishGoToItem = [this, weakThis = WeakPtr { this }, frameLoadType, shouldTreatAsContinuingLoad, page, targetItem = Ref { targetItem }] (bool result) {
```

Comment:
We should not be capturing both `this` and and `weakThis` per Safer CPP guidelines. It's unfortunate the Safer CPP bot ignores this file for now. We should:
1. Capture `weakThis` only, not `this`
2. Convert the `weakThis` into a `protectedThis` at the beginning of the lambda
3. Use `protectedThis->` whenever necessary (if the logic is too long and that's too many `protectedThis->`, then it indicates we should move that logic to a separate member function.

Reply:
https://github.com/WebKit/WebKit/pull/40319

Reply:
As established in comments here and on slack, I went the weak route because the straightforward ref/deref forwarding broke the build with static asserts.

We found the workaround for that.